catalog/core/catalog-core-metricsplugin/src/main/java/ddf/catalog/metrics/CatalogMetrics.java

```diff
@@ -13,226 +13,288 @@
+    addStartTime(input);
     return input;
   }
 
   // PostDelete
   @Override
   public DeleteResponse process(DeleteResponse input) throws PluginExecutionException {
-    if (Requests.isLocal(input.getRequest())) {
-      deletedMetacards.mark(input.getDeletedMetacards().size());
-    }
+    recordLatency(input, DELETE_SCOPE);
+    deletedMetacards.increment(input.getDeletedMetacards().size());
+    recordExceptions(input.getProcessingErrors(), DELETE_SCOPE);
+
+    return input;
+  }
+
+  // PreResource
+  @Override
+  public ResourceRequest process(ResourceRequest input)
+      throws PluginExecutionException, StopProcessingException {
+    addStartTime(input);
     return input;
   }
 
   // PostResource
   @Override
   public ResourceResponse process(ResourceResponse input)
       throws PluginExecutionException, StopProcessingException {
-    resourceRetrival.mark();
+    recordLatency(input, RESOURCE_SCOPE);
+    recordExceptions(input.getProcessingErrors(), RESOURCE_SCOPE);
+
     return input;
   }
 
-  private void recordSourceQueryExceptions(QueryResponse response) {
-    Set<ProcessingDetails> processingDetails =
-        (Set<ProcessingDetails>) response.getProcessingDetails();
-
-    if (processingDetails == null || processingDetails.iterator() == null) {
+  private void recordExceptions(Set<ProcessingDetails> processingDetails, String operation) {
+    if (processingDetails == null) {
       return;
     }
 
-    Iterator<ProcessingDetails> iterator = processingDetails.iterator();
-    while (iterator.hasNext()) {
-      ProcessingDetails next = iterator.next();
+    for (ProcessingDetails next : processingDetails) {
       if (next != null && next.getException() != null) {
-        if (next.getException() instanceof UnsupportedQueryException) {
-          unsupportedQueryExceptions.mark();
-        } else if (next.getException() instanceof SourceUnavailableException) {
-          sourceUnavailableExceptions.mark();
-        } else if (next.getException() instanceof FederationException) {
-          federationExceptions.mark();
-        }
-        exceptions.mark();
+        String exceptionName = rootCauseExceptionName(next.getException());
+        Metrics.counter(
+                metricName(METRIC_PREFIX, operation, EXCEPTIONS_SCOPE),
+                "type",
+                exceptionName,
+                "source",
+                next.getSourceId())
+            .increment();
       }
     }
+  }
+
+  private String rootCauseExceptionName(Exception exception) {
+    Throwable rootCause = exception;
+    while (rootCause.getCause() != null && rootCause.getCause() != rootCause) {
+      rootCause = rootCause.getCause();
+    }
+    return rootCause.getClass().getName();
+  }
+
+  private String metricName(String... parts) {
+    return String.join(".", parts);
+  }
+
+  private Set<String> getSourceIds(QueryRequest query) {
+    Set<String> result;
+
+    if (query.isEnterprise()) {
```

Comment:
❓ Why not just immediately return when the predicate is valid instead of doing assignment?

```
    if (query.isEnterprise())
      return Collections.singleton("enterprise");

    if (query.getSourceIds() == null || query.getSourceIds().isEmpty()) {
      return Collections.singleton(SystemInfo.getSiteName());

    return query.getSourceIds();
```

Reply:
https://stackoverflow.com/a/733858

This method is small enough so I do not think it matters either way in the context of the Code Complete advice.